kratos/input_output/gid_output.cpp

```diff
@@ -0,0 +1,1437 @@
+    const int ValueIndex
+    )
+{
+    KRATOS_TRY;
+
+    Timer::Start("Writing Results");
+
+    for ( auto it = mGidGaussPointContainers.begin(); it != mGidGaussPointContainers.end(); it++ ) {
+        it->PrintResults( mResultFile, rVariable, rModelPart, SolutionTag, ValueIndex );
+    }
+
+    Timer::Stop("Writing Results");
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TGaussPointContainer, class TMeshContainer>
+void GidIO<TGaussPointContainer, TMeshContainer>::PrintOnGaussPoints(
+    const Variable<array_1d<double,3> >& rVariable,
+    const ModelPart& rModelPart,
+    const double SolutionTag,
+    const int ValueIndex
+    )
+{
+    KRATOS_TRY;
+
+    Timer::Start("Writing Results");
+
+    for ( auto it = mGidGaussPointContainers.begin(); it != mGidGaussPointContainers.end(); it++ ) {
+        it->PrintResults(  mResultFile, rVariable, rModelPart, SolutionTag, ValueIndex );
+    }
+
+    Timer::Stop("Writing Results");
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TGaussPointContainer, class TMeshContainer>
+void GidIO<TGaussPointContainer, TMeshContainer>::PrintOnGaussPoints(
+    const Variable<Vector>& rVariable,
+    const ModelPart& rModelPart,
+    const double SolutionTag,
+    const int ValueIndex
+    )
+{
+    KRATOS_TRY;
+    Timer::Start("Writing Results");
+
+    for ( auto it = mGidGaussPointContainers.begin(); it != mGidGaussPointContainers.end(); it++ ) {
+        it->PrintResults(  mResultFile, rVariable, rModelPart, SolutionTag, ValueIndex );
+    }
+
+    Timer::Stop("Writing Results");
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<class TGaussPointContainer, class TMeshContainer>
+void GidIO<TGaussPointContainer, TMeshContainer>::PrintOnGaussPoints(
+    const Variable<Matrix>& rVariable,
+    const ModelPart& rModelPart,
+    const double SolutionTag,
+    const int ValueIndex
+    )
+{
+    KRATOS_TRY;
+    Timer::Start("Writing Results");
+
+    for ( auto it = mGidGaussPointContainers.begin(); it != mGidGaussPointContainers.end(); it++ ) {
+        it->PrintResults(  mResultFile, rVariable, rModelPart, SolutionTag, ValueIndex );
+    }
+
+    Timer::Stop("Writing Results");
+
+    KRATOS_CATCH("");
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+// GidIO default instantiation
+template class GidIO<GidGaussPointsContainer,GidMeshContainer>;
```

Comment:
I am not sure if you can split header/source
You basically prevent ppl from having their own derived versions with different templates
@roigcarlo what do you think?

Reply:
Can you please try to compile the `FreeSurfaceApplication` to check this? (they have their own derived version with an edge based Gauss point container).

Reply:
This needs to be checked both in Win and Linux
maybe add this app temporarily to the CI to check

Reply:
I did the changes done in the FreeSurfaceApplication

Reply:
I think you misunderstand
Just try to compile the App

It should not work because it misses the instantiation with the types from the FreeSurfaceApp

Reply:
OK

Reply:
I cannot...

Reply:
I cannot means you can compile it or means that the application does not compile?

Reply:
The application does not compile, so or I move the source to the header again or I move the FreeSurface gid containers to the core...

Reply:
More the source to the header again.

Reply:
Yes, but the point is to reduce compilation times of including gid_io

Reply:
Hum, it should compile as long as you expose both the declaration and the instantiation.

Reply:
OK;  i Will see

Reply:
I mean you should add something like:

`KRATOS_API_EXTERN template class KRATOS_API(KRATOS_CORE) GidIO<GidGaussPointsContainer,GidMeshContainer>;`

to gid_io.h, outside the class

Reply:
I was doing that

Reply:
:)

Reply:
I fixed it with a little bit of triquery